python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,445 @@
+"""Definitions of types related to coordinate systems."""
+
+import abc
+import collections.abc
+from typing import Optional, Sequence, Tuple, Union
+
+import attrs
+import numpy as np
+import numpy.typing as npt
+
+from .types import str_or_seq_length
+from .types import to_string_tuple
+
+
+@attrs.define(frozen=True)
+class Axis:
+    """A description of an axis of a coordinate system
+
+    Lifecycle: experimental
+    """
+
+    name: str
+    """Name of the axis."""
+    unit: Optional[str] = None
+    """Optional string name for the units of the axis."""
+
+
+@attrs.define(frozen=True)
+class CoordinateSpace(collections.abc.Sequence):
+    """A coordinate space for spatial data.
+
+    Args:
+        axes: The axes of the coordinate system in order.
+
+    Lifecycle: experimental
+    """
+
+    axes: Tuple[Axis, ...] = attrs.field(converter=tuple)
+
+    @axes.validator
+    def _validate(self, _, axes: Tuple[Axis, ...]) -> None:
+        if not axes:
+            raise ValueError("at least one")
```

Comment:
whoops, the strings I used in my example above were just short references to what the full message should be; these will be raised as-is, so they should have the full description.

Reply:
... I'm embarrassed I didn't catch that when I copied it over.